docs/blueprints/zombie-segments-cleaner.md

```diff
@@ -0,0 +1,57 @@
+# Zombie Segments Cleaner
+
+## Abstract
+
+This document describes a design for cleaning segments that are not accessible with standard operations like listing.
+
+## Background
+
+Currently with a multisegment object, we have several places where we can leave unfinished or broken objects:
+* if upload process will be stopped, because of an error or canceling, when one or more segments are already uploaded we can become zombie segments.
+* if delete process will be stopped, because of an error or canceling, when one or more segments are already deleted we can have an object with an incomplete number of segments.
+* in the past we had a bug where error during deleting segments from nodes was interrupting deleting segment from satellite, and we have segments available on satellite but not on storage node.
+* deleted bucket can leave all objects inaccessible ???
+* deleted project can leave all objects inaccessible ???
+
+We need a system for identifying objects with missing segments, and zombie segments that are not listed as part of objects, that cleans up those segments.
+
+In the long term, the metainfo refactor will fix this. We need a short term solution.
+
+## Design
+
+How detect zombie segment:
+* segment doesn't have corresponding last segment (`l`)
+**Note**: Exception from this case are segments that are part of object that is currently uploaded. For time of uploading all object segments have indexes `s0..sN` but at the end of upload (`CommitObject`) segment `sN` is renamed (`delete` and `put`) to last segment (`l`).
+* segment where index is greater than 0 but any previous segment is missing
+* last segment (`l`) where unencrypted number of segments is greater than 0 but rest of segments are missing or number of existing segments is different from stored value
+
+Different kind of bad segment is a case where all segments are available on satellite but were deleted from storage nodes.
+
+General idea is to create cli command to verify all segment paths and delete bad segments. With this command user should be able to specify flags like DB connection string, dry run (only listing bad segments) and how old segments should be verified. During iteration, each segment path will be processed and assigned to a struct that will collect segments from the same object. All segments from object where at least one segment is not old enough should be skipped. When the last segment (`l`) will be reached or iteration will be finished each object will be checked if it contains all segments (according to rules from the beginning of this part). If yes, then helper struct will be removed from memory. If not, then all related segments will be moved to delete or only printed in case of dry run.
+
+Command can be run agains production database or in case of performance concerns agains backup of pointerDB.
```

Comment:
I'd suggest having 2 separate CLI tools:
- zombie detector
- zombie repear

The zombie detector tool will scan the database and will build the list of zombie segments to be deleted. The list will include the segment path and the segment create/modified date. We should run the detector on a snapshot copy of the production database, not on the live production database. Not only for performance reasons, but also for generating more reliable reports. It would be harder to implement zombie detection if the database content is changing while iterating over it.

On the other hand, we will execute the zombie reaper tool on the live production database. It will take the report from the zombie detector tool and will delete the corresponding segments. Before deletion, it will check if the create/modified is still the same. If it is not the same, this means that the segment has been reuploaded since the detection phase and it should not be deleted. We should implement this conditional deletion in a compare-and-swap operation. It would still make sense to have an option for a dry run.

Reply:
I think its a very good idea. 

> We should implement this conditional deletion in a compare-and-swap operation.

Can you elaborate more about this statement? I'm not sure I fully understand.

Reply:
The conditional deletion should have 2 logical steps:
1. Check the create/modified date
2. Delete the segment if the check from the previous step passes

The above 2 steps should be implemented in an atomic operation to avoid race conditions. Imagine, the segment's pointer is overwritten due to re-upload just in the gap between steps 1 and 2. In such a case, we would delete the newly uploaded segment instead of the supposed zombie segment. We will lose data!

One way to implement the 2 steps in an atomic operation is to use compare-and-swap. For example, we use this for atomically update the pieces in the pointer: https://github.com/storj/storj/blob/0885eaa183571896da373f7e782cae8d67160069/satellite/metainfo/service.go#L152

In the case of deleting segments, it should be like:
```go
s.DB.CompareAndSwap(ctx, []byte(path), oldPointerBytes, nil)
```
This should ensure that the pointer has not changed since we read it for checking the create/modified date, and we are not deleting a newer pointer.

Reply:
I see but it means that we need to have serialized pointer in `detect` command report, right?

Reply:
No. The `detect` command should include only the path and the creation date, not the complete serialized pointer.

The serialized pointer will come from a normal `Get` from the DB:

```go
// read the pointer
pointerBytes, err := s.DB.Get(ctx, []byte(path))
if err != nil {
	return err
}
// unmarshal the pointer
pointer = &pb.Pointer{}
err = proto.Unmarshal(pointerBytes, pointer)
if err != nil {
	return err
}
// check if pointer has been replaced
if !pointer.GetCreationDate().Equal(creationDateFromReport) {
        // pointer has been replaced since detection, do not delete it.
	return nil
}
// delete the pointer using compare-and-swap
err = s.DB.CompareAndSwap(ctx, []byte(path), pointerBytes, nil)
if storage.ErrValueChanged.Has(err) {
	// race detected while deleting the pointer, do not try deleting it again.
	return nil
}
if err != nil {
	return err
}

Reply:
Ah right, thanks